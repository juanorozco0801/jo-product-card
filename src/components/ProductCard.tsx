import React, { createContext, CSSProperties } from 'react';

import { useProduct } from '../hooks/useProduct'
import { IInitialValues, IOnChangeArgs, IProduct, IProductCardHandlers, IProductContextProps } from '../interfaces/interfaces';


import styles from '../styles/styles.module.css'


export const ProductContext = createContext({} as IProductContextProps);
const { Provider } = ProductContext;

export interface IProps {
  product: IProduct;
  // children?: ReactElement | ReactElement[];
  children: (args:IProductCardHandlers) => JSX.Element;
  className?: string;
  style?: CSSProperties;
  onChange?: ( args: IOnChangeArgs) => void;
  value?: number;
  initialValues?: IInitialValues;
}

export const ProductCard = ({ product, children, className, style, onChange, value, initialValues }: IProps) => {

  const { counter, increasedBy, maxCount, isMaxCountReached, reset } = useProduct( {
    onChange, product, value, initialValues
  });


  return (
    <Provider value={{
      counter,
      increasedBy,
      maxCount,
      product,
    }}>

      <div
        className={`${styles.productCard} ${className}`}
        style={style}
      >

        { children({
          counter,
          increasedBy,
          product,
          maxCount,
          isMaxCountReached,
          reset,
        }) }

      </div>
    </Provider>
  )
};
