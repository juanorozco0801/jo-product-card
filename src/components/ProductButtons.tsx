import React, { CSSProperties, useCallback, useContext } from 'react';

import { ProductContext } from "./ProductCard";

import styles from '../styles/styles.module.css'

export interface IProps {
  className?: string
  style?: CSSProperties
}

export const ProductButtons = ({ className, style }: IProps ) => {

  const { increasedBy, counter, maxCount } = useContext( ProductContext )

  const isMaxReached = useCallback(
    () => !!maxCount && counter === maxCount,
    [counter, maxCount],
  )
  

  return(
    <div
      className={`${styles.buttonsContainer} ${className}`}
      style={style}  
    >
    <button
      className={styles.buttonMinus}
      onClick={() => increasedBy(-1)}
    >
      -
    </button>
    <div className={styles.countLabel}>{counter}</div>
    <button
      className={`${styles.buttonAdd} ${isMaxReached() && styles.disabled}`}
      onClick={() => increasedBy(1)}
    >
      +
    </button>
  </div>
  )
}