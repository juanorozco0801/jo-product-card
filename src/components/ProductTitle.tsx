import React, { CSSProperties, useContext } from 'react';

import { ProductContext } from "./ProductCard";

import styles from '../styles/styles.module.css'

export interface IProps {
  className?: string
  name?: string,
  style?: CSSProperties
}

export const ProductTitle = ({ name, className, style }: IProps) => {

  const { product } = useContext(ProductContext)

  return (
    <span
      className={`${styles.productDescription} ${className}`}
      style={style}
    >
      { name ? name : product.name }
    </span>
  )
}