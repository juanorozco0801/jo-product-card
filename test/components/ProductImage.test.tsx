import React from 'react';
import renderer from 'react-test-renderer';
import { ProductCard } from "../../src/components/ProductCard";
import { ProductImage } from "../../src/components/ProductImage";
import { product2 } from "../data/product";

describe('ProductImage', () => {

  it('Should display the component correctly with the image sent', () => {

    const wrapper = renderer.create(
      <ProductImage img="custom-image.jpg" className='custom-class'/>
    )

    expect( wrapper.toJSON() ).toMatchSnapshot();

  });

  it ('should display the component correctly with the image in the product', () => {

      const wrapper =  renderer.create(
        <ProductCard product={ product2 }>
          {
            () => (
              <ProductImage />
            )
          }
  
        </ProductCard>
      )
  
      expect( wrapper.toJSON() ).toMatchSnapshot();

  });

});