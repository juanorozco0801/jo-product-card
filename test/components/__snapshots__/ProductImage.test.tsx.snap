// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage Should display the component correctly with the image sent 1`] = `
<img
  alt="Coffee mug"
  className="productImg custom-class"
  src="custom-image.jpg"
/>
`;

exports[`ProductImage should display the component correctly with the image in the product 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Coffee mug"
    className="productImg undefined"
    src="./coffee-mug.png"
  />
</div>
`;
